### PR TITLE
Replace `dm-tree` with `optree`

### PR DESCRIPTION
**Pull request by james77777778**
Related to #18442 
Related to #18614 

This PR refactors `keras.utils.tree` to use `optree` instead of `dm-tree`

- Exports APIs (`is_nested`, `flatten`, `unflatten_as`, `map_structure`, `map_structure_up_to`, `assert_same_structure`, `pack_sequence_as`, `lists_to_tuples`) with the path `keras.utils.tree.*`
- Add docstrings (mostly borrowed from `dm-tree` and `tf.nest`)
- Adds unit tests
- Eliminates `tf.nest` in the codebase (excluding legacy code)

I have verified that exported APIs should meet the requirements of keras_cv and keras_nlp

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR. Do you observe performance difference? How long does it take to run the unit test suite with the PyTorch backend before and after the change, for instance?

**Comment by james77777778** [comment]
> Thanks for the PR. Do you observe performance difference? How long does it take to run the unit test suite with the PyTorch backend before and after the change, for instance?

I didn't observe a performance difference in the unit tests.
Therefore, I benchmarked the performance using the method in #18569 

|backend|cpu/cuda|`jit_compile`|master branch|optree|
|-|-|-|-|-|
|torch|cpu|`False`|12~13ms|11~12ms|
|||`True`|20ms|17~19ms|
||cuda|`False`|2~3ms|3ms|
|||`True`|5~6ms|6ms|

There is no difference observed when using cuda, but there is a slight improvement when using cpu
However, even with `jit_compile=True`, it still runs slower compared to when `jit_compile=False`

File: keras/utils/tracking.py [hunk 1]
```diff
@@ -110,6 +113,7 @@ def add_to_store(self, store_name, value):
         self.stored_ids[store_name].add(id(value))
 
 
+@optree.register_pytree_node_class(namespace="keras")
```
Comment (fchollet) [inline keras/utils/tracking.py:116]: Why is this necessary?

File: keras/ops/core_test.py [hunk 2]
```diff
@@ -790,8 +790,8 @@ def test_cond_check_output_spec_list_tuple(self):
     def test_cond_check_output_spec_other_types(self):
         cond_op = core.Cond()
         # Create mock objects with dtype and shape attributes
```
Comment (fchollet) [inline keras/ops/core_test.py:792]: You can remove this comment.

File: keras/utils/tree.py [hunk 3]
```diff
@@ -1,60 +1,331 @@
...
+
+    optree.register_pytree_node(
+        ListWrapper,
+        lambda x: (x, None),
+        lambda metadata, children: ListWrapper(list(children)),
+        namespace="keras",
+    )
+
+
+@keras_export("keras.utils.tree.is_nested")
```
Comment (fchollet) [inline keras/utils/tree.py:32]: What is the case for exporting these utils to the public API?

File: keras/utils/tracking.py (hunk 1 above)
Comment (james77777778) [reply keras/utils/tracking.py:116]: Because we need tree manipulations on `TrackedList`, `TrackedDict` and `TrackedSet` (as well as `ListWrapper`) and `optree` requires explicit registration

Please refer to https://github.com/metaopt/optree?tab=readme-ov-file#registering-a-container-like-custom-type-as-non-leaf-nodes

File: keras/utils/tree.py (hunk 3 above)
Comment (james77777778) [reply keras/utils/tree.py:22]: Kindly ping @mattdangerw as mentioned in https://github.com/keras-team/keras/issues/18614#issuecomment-1764960732
> If we add a `keras.utils.tree` let's make sure to export it.

I have confirmed that:
keras_cv requires `tree.flatten` and `tree.map_structure` (https://github.com/search?q=repo%3Akeras-team%2Fkeras-cv%20tree.&type=code)
keras_nlp also requires `tree.flatten` and `tree.map_strcture` (https://github.com/search?q=repo%3Akeras-team%2Fkeras-nlp%20tree.&type=code)

Exporting these 2 apis may be sufficient?

File: keras/utils/tree.py [hunk 4]
```diff
@@ -88,55 +359,148 @@ def truncate(value, length):
...
+        elif isinstance(instance, types.MappingProxyType):
+            # MappingProxyType requires a dict to proxy to.
+            return type(instance)(dict(keys_and_values))
+        else:
+            return type(instance)(keys_and_values)
+    elif isinstance(instance, collections.abc.MappingView):
+        # We can't directly construct mapping views, so we create a list instead
+        return list(args)
+    elif optree.is_namedtuple(instance):
+        if isinstance(instance, ObjectProxy):
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras/utils/tree.py]: Done. The tests passed.

File: keras/utils/tree.py (hunk 3 above)
Comment (fchollet) [reply keras/utils/tree.py:22]: If we have these 2 we might as well export everything. I'd recommend moving them to `keras.tree`.

File: keras/utils/tracking.py (hunk 1 above)
Comment (fchollet) [reply keras/utils/tracking.py:116]: Got it, that's pretty inconvenient but that makes sense.

File: keras/utils/tree.py (hunk 3 above)
Comment (james77777778) [reply keras/utils/tree.py]: Done. Exports APIs using `keras.tree.*`

**Review by fchollet — APPROVED** [review]
LGTM, thank you for the contribution!
